### DISCUSSION TITLE
[Proposal] Implicit Generic Types

### DISCUSSION THREAD

**Discussion by PereViader** [dotnet/csharplang#5730] [Language Ideas]
# Introduction

When writing Generic classes/interfaces/methods, all the types have to be explicitly defined on the signature in order to be able to assert everything properly follows the type constraints.

This however adds some unnecessary noise when referencing those types.

As an example, given these types:

```
interface IInterface<T> { }

class SomeClass : IInterface<int> { }
```

and this method signature:

```
void Method<T, Y>(Y arg) where T : IInterface<Y> { }
```

Notice that in order to call this method you have to call it like this

```
Method<SomeClass, int>(42);
```

There is an opportunity to improve the type system and make this more concise by detecting the `int` type implicitly from `SomeClass`

The expected way to call this method after implementing this feature would be:

```
Method<SomeClass>(42);
```

# Implementation proposal

## Proposal 1

Add a new area where to place implicit types after the explicit types
These types must be then used to constraint other types on the `where`

```
void Method<T><Y>(Y arg) where T : IInterface<Y> { }
```

## Proposal 2

Add a new area where to place implicit types before the where clauses
These types must be then used to constraint other types on the `where`

```
void Method<T>(Y arg) 
    implicit Y
    where T : IInterface<Y> { }
```

## Proposal 3

Be able to define types with the `implicit` keyword. 
These types must be then used to constraint other types on the `where`

```
void Method<T, implicit Y>(Y arg) 
    where T : IInterface<Y> { }
```

--- 
Edit1:
# Further Examples

Given the previous types and this method for the current c# version.

```
Y Method2<T, Y>(T arg) where T : IInterface<Y>
{
    return default;
}
```

You would call this method as

```
SomeClass arg = new ();
int value = Method2<SomeClass, int>(arg);
```

When it could be more concisely called as

```
SomeClass arg = new ();
int value = Method2(arg);
```

As all the types could be implicitly resolved by the type system.

**Comment by 333fred** [comment]
This looks like #5556.

**Reply by PereViader** [reply]
I think they serve a bit different purposes.

**Reply by ufcpp** [reply]
https://github.com/dotnet/csharplang/issues/1349

**Reply by PereViader** [reply]
@ufcpp This is defenetly related to that proposal.
Proposals are different however, so I don't know if this post has should be kept alive.
Their proposals ask the type system to be smarter with the same signature syntax.
This proposal asks for an extension of the generic syntax to make the implicit nature of some types explicit.

**Reply by 333fred** [reply]
> I think they serve a bit different purposes.

I can't say I agree. This is specifically about hiding type parameters by way of "type 1 implies type 2". That's exactly what associated/existential types gets you.

**Reply by PereViader** [reply]
@333fred Not really an expert on this. Let me know if I need to do anything.

**Comment by PereViader** [comment]
Maybe related to https://github.com/dotnet/csharplang/discussions/129
